v3/src/components/graph/components/dotplotdots.tsx

```diff
@@ -1,102 +1,207 @@
+        const numBins = Math.ceil(plotWidth / defaultDiameter) + 1,
+          binWidth = plotWidth / (numBins - 1),
+          bins: string[][] = range(numBins + 1).map(() => [])
+
+        dataset?.cases.forEach(({ __id__ }) => {
+          const numerator = xScale?.(dataset?.getNumeric(__id__, xAttrID) ?? -1),
+            bin = Math.ceil((numerator ?? 0) / binWidth)
+          if (bin >= 0 && bin <= numBins) {
+            bins[bin].push(__id__)
+            binMap[__id__] = {yIndex: bins[bin].length}
+          }
+        })
+        const maxInBin = (max(bins, (b => b.length)) || 0) + 1,
+          excessHeight = Math.max(0, maxInBin - Math.floor(yHeight / defaultDiameter)) * defaultDiameter
+        overlap = excessHeight / maxInBin
+      }
+
+      computeBinPlacements()
+
+      const
+        getScreenX = (anID: string) => getScreenCoord(dataset, anID, xAttrID, xScale),
+        computeYCoord = (binContents: { yIndex: number }) => {
+          return binContents ? yHeight - defaultRadius / 2 - binContents.yIndex * (defaultDiameter - overlap) : 0
+        },
+        getScreenY = (anID: string) => computeYCoord(binMap[anID]),
+        duration = firstTime.current ? transitionDuration : 0,
+        onComplete = firstTime.current ? () => {
+          prf.measure("Graph.ScatterDots[refreshPointPositions]", () => {
+            firstTime.current = false
+          })
+        } : undefined
+
+      setPointCoordinates({dotsRef, selectedOnly, getScreenX, getScreenY, duration, onComplete})
+    })
+  }, [dataset, dotsRef, xAttrID, xScale, yScale, plotWidth])
+
+  // respond to axis domain changes (e.g. axis dragging)
+  useEffect(() => {
+    refreshPointPositions(false)
+    const disposer = reaction(
+      () => [axisModel.domain],
+      domains => {
+        firstTime.current = false // don't animate response to axis changes
+        refreshPointPositions(false)
+      }
+    )
+    return () => disposer()
+  }, [refreshPointPositions, axisModel.domain])
+
+  // respond to axis range changes (e.g. component resizing)
+  useEffect(() => {
+    refreshPointPositions(false)
+    const disposer = reaction(
+      () => [layout.axisLength(place), layout.axisLength(countPlace)],
+      ranges => {
+        firstTime.current = false // don't animate response to axis changes
+        refreshPointPositions(false)
+      }
+    )
+    return () => disposer()
+  }, [layout, refreshPointPositions, place, countPlace])
+
+  // respond to selection and value changes
+  useEffect(() => {
+    const disposer = dataset && onAction(dataset, action => {
+      if (isSelectionAction(action)) {
+        refreshPointSelection()
+      } else if (isSetCaseValuesAction(action)) {
+        // assumes that if we're caching then only selected cases are being updated
+        refreshPointPositions(dataset.isCaching)
+      }
+    }, true)
+    return () => disposer?.()
+  }, [dataset, refreshPointPositions, refreshPointSelection])
```

Comment:
Not necessary for this PR, but these three `useEffects()` are now virtually identical between the `DotPlotDots` and `ScatterDots`, which suggests that there's the potential for a reusable custom hook that could be used by both.

Reply:
I agree. Thinking the same myself.